cpp/src/arrow/compute/kernels/aggregate_basic_sse.cc

```diff
@@ -0,0 +1,355 @@
+#include "arrow/compute/api_aggregate.h"
+#include "arrow/compute/kernels/aggregate_basic_internal.h"
+#include "arrow/compute/kernels/aggregate_internal.h"
+#include "arrow/compute/kernels/common.h"
+#include "arrow/util/align_util.h"
+#include "arrow/util/simd.h"
+
+TARGET_CODE_START_SSE4_2
+namespace arrow {
+namespace compute {
+namespace aggregate {
+
+// ----------------------------------------------------------------------
+// Sum implementation for SSE
+
+// Each m128 stream handle 2 double/int64 accumulator type, one batch has 4 streams.
+static constexpr int kSseBatchStreams = 4;
+static constexpr int kSseStreamSize = sizeof(__m128d) / sizeof(double);
+static constexpr int kSseBatchSize = kSseBatchStreams * kSseStreamSize;
+
+// Default scalar version
+template <typename T, typename SumT>
+inline SumResult<SumT> SumDenseBatchSse(const T* values, int64_t num_batch) {
+  SumResult<SumT> sum_result;
+  SumT sum_streams[kSseBatchSize] = {0};
+
+  // Add the results by streams
+  for (int64_t batch = 0; batch < num_batch; batch++) {
+    for (int i = 0; i < kSseBatchSize; i++) {
+      sum_streams[i] += values[(batch * kSseBatchSize) + i];
+    }
+  }
+
+  // Aggregate the result streams
+  for (int i = 0; i < kSseBatchSize; i++) {
+    sum_result.sum += sum_streams[i];
+  }
+  sum_result.count = num_batch * kSseBatchSize;
+  return sum_result;
+}
+
+// Dense helper for accumulator type is same to data type
+#define SUM_DENSE_BATCH_SSE_DIRECT(Type, SumSimdType, SimdZeroFn, SimdLoadFn, SimdAddFn) \
+  template <>                                                                            \
+  inline SumResult<Type> SumDenseBatchSse(const Type* values, int64_t num_batch) {       \
+    SumResult<Type> sum_result;                                                          \
+    SumSimdType results_simd[kSseBatchStreams];                                          \
+    for (int i = 0; i < kSseBatchStreams; i++) {                                         \
+      results_simd[i] = SimdZeroFn();                                                    \
+    }                                                                                    \
+                                                                                         \
+    /* Add the values to result streams */                                               \
+    for (int64_t batch = 0; batch < num_batch; batch++) {                                \
+      for (int i = 0; i < kSseBatchStreams; i++) {                                       \
+        const auto src_simd =                                                            \
+            SimdLoadFn(&values[batch * kSseBatchSize + kSseStreamSize * i]);             \
+        results_simd[i] = SimdAddFn(src_simd, results_simd[i]);                          \
+      }                                                                                  \
+    }                                                                                    \
```

Comment:
Have you considered using OpenMP SIMD (i.e., `-fopenmp-simd`, which does not do anything with threads or require linking OpenMP)?
```c
double sum1(int64_t n, const double *values) {
    double sum = 0;
    #pragma omp simd reduction(+:sum)
    for (int64_t i = 0; i < n; i++)
        sum += values[i];
    return sum;
}
```
produces good vectorized code (probably excessively unrolled by clang)
```
        vaddpd  zmm0, zmm0, zmmword ptr [rsi + 8*rdx]
        vaddpd  zmm1, zmm1, zmmword ptr [rsi + 8*rdx + 64]
        vaddpd  zmm2, zmm2, zmmword ptr [rsi + 8*rdx + 128]
        vaddpd  zmm3, zmm3, zmmword ptr [rsi + 8*rdx + 192]
        vaddpd  zmm0, zmm0, zmmword ptr [rsi + 8*rdx + 256]
        vaddpd  zmm1, zmm1, zmmword ptr [rsi + 8*rdx + 320]
        vaddpd  zmm2, zmm2, zmmword ptr [rsi + 8*rdx + 384]
        vaddpd  zmm3, zmm3, zmmword ptr [rsi + 8*rdx + 448]
        vaddpd  zmm0, zmm0, zmmword ptr [rsi + 8*rdx + 512]
        vaddpd  zmm1, zmm1, zmmword ptr [rsi + 8*rdx + 576]
        vaddpd  zmm2, zmm2, zmmword ptr [rsi + 8*rdx + 640]
        vaddpd  zmm3, zmm3, zmmword ptr [rsi + 8*rdx + 704]
        vaddpd  zmm0, zmm0, zmmword ptr [rsi + 8*rdx + 768]
        vaddpd  zmm1, zmm1, zmmword ptr [rsi + 8*rdx + 832]
        vaddpd  zmm2, zmm2, zmmword ptr [rsi + 8*rdx + 896]
        vaddpd  zmm3, zmm3, zmmword ptr [rsi + 8*rdx + 960]
        .loc    1 6 5                   # ./example.c:6:5
        sub     rdx, -128
        add     rcx, -4
        jne     .LBB0_7
```
https://gcc.godbolt.org/z/6mD6oy

Reply:
Thanks, the vectorize code by clang is really good. But seems the GCC part only use zmm0 which may limit the performance. This PR only aims on dense part, the final target is expanding the support to sparse part also, do you know if sparse can applied similar approach also? Example scalar code like below

  inline T MaskedValue(bool valid, T value) const { return valid ? value : 0; }
      for (size_t i = 0; i < 8; i++) {
        local.sum += MaskedValue(bits & (1U << i), values[i]);
      }
      local.count += BitUtil::kBytePopcount[bits];

Reply:
You can use GCC pragmas to force more unrolling, but AVX512 vectorization gets you a ways despite the adder instruction latency. (My experience has been that GCC does an excellent job with intrinsics, often better than ICC, and tends to favor smaller code at comparable optimization levels. Clang tends to be more aggressive than GCC at unrolling, though I've often seen that not pay off.)

The masked case is harder and I'm not aware of an idiomatic way to get GCC to vectorize it. We get some masked vectorized code with clang, though it isn't as clean as I'd hope for.
```c
double sum1(int64_t n, const unsigned char *mask, const double *values) {
    double sum = 0;
    #pragma omp simd reduction(+:sum) collapse(2)
    for (int64_t i = 0; i < n/8; i++) {
        for (int64_t j = 0; j < 8; j++) {
            sum += mask[i] & (1 << j) ? values[i*8+j] : 0;
        }
    }
    return sum;
}
```

https://gcc.godbolt.org/z/84anmr